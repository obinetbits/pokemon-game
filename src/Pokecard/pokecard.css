.flip-card {
	background-color: transparent;
	perspective: 1000px; /* Remove this if you don't want the 3D effect */
}
/* This container is needed to position the front and back side */
.flip-card-inner {
	position: relative;
	text-align: center;
	transition: transform 0.8s;
	transform-style: preserve-3d;
}

/* Do an horizontal flip when you move the mouse over the flip box container */
.flip-card-action .flip-card-inner {
	transform: rotateY(180deg);
}

.Pokecard-front,
.Pokecard-back {
	width: 200px;
	padding: 10px;
	margin: 1rem;
	box-shadow: 7px 10px 12px -5px rgba(0, 0, 0, 0.56);
	text-align: center;
	background-color: #fff;
	border-radius: 3px;
	backface-visibility: hidden;
}

.Pokecard-data {
	font-size: 80%;
}

.Pokecard-title {
	margin-bottom: 1rem;
	font-weight: bold;
	font-size: 1.5rem;
}

.Pokecard-front {
	position: absolute;
	cursor: pointer;
}

.Pokecard-back {
	transform: rotateY(180deg);
}
