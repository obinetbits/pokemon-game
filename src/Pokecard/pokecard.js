import React, { useCallback } from 'react';
import './pokecard.css';

// https://media1.tenor.com/images/84c5b716f0f747acc57a8176e3e6affd/tenor.gif?itemid=4444793
// https://media1.tenor.com/images/5fa6f43961a2c4e5e08ea014bd958459/tenor.gif?itemid=5964096
// https://media1.tenor.com/images/5fa6f43961a2c4e5e08ea014bd958459/tenor.gif'
// https://media1.tenor.com/images/84c5b716f0f747acc57a8176e3e6affd/tenor.gif

// const POKE_API = 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/';
const POKE_API_I = 'https://assets.pokemon.com/assets/cms2/img/pokedex/detail/';
const POKE_API_II = 'http://www.pokestadium.com/sprites/xy/'; // animate

const padToThree = (num) => (num <= 999 ? `00${num}`.slice(-3) : num);

const Pokecard = ({
	wasJustClickedOn,
	flipCard,
	uniqueId,
	name,
	imageId,
	isCurrentName,
	isPrevName,
	isMatched,
	exp,
	type
}) => {
	const handleFlipCard = useCallback(
		() => {
			if (wasJustClickedOn) return;
			flipCard({ uniqueId, name });
		},
		[ wasJustClickedOn, flipCard, uniqueId, name ]
	);

	const handleImageError = useCallback((e) => {
		console.log({ e });
		e.target.src = `https://media1.tenor.com/images/5fa6f43961a2c4e5e08ea014bd958459/tenor.gif`;
	}, []);

	// const imgSrc = `${POKE_API_I}${padToThree(imageId)}.png`;
	const imgSrc = `${POKE_API_II}${name}.gif`;
	const stayOpen = isCurrentName && isPrevName;
	const showCard = wasJustClickedOn || stayOpen || isMatched;
	return (
		<div className={`flip-card${showCard ? ' flip-card-action' : ''}`}>
			<div className="flip-card-inner">
				<div className="Pokecard-front" onClick={handleFlipCard}>
					<h1 className="Pokecard-title">Guess</h1>
					<div className="Pokecard-image">
						<h1>Pokemon Card</h1>
					</div>
					<div className="Pokecard-data">
						<span role="img" aria-label="Monser">
							👾👾👾
						</span>
					</div>
					<div className="Pokecard-data">
						<span role="img" aria-label="Diamonds">
							♦️♦️️♦️♦️
						</span>
					</div>
				</div>
				<div className="Pokecard-back">
					{showCard ? (
						<React.Fragment>
							<h1 className="Pokecard-title">{name}</h1>
							<div className="Pokecard-image">
								<img src={imgSrc} onError={handleImageError} alt={name} />
							</div>
							<div className="Pokecard-data">Exp: {exp} </div>
							<div className="Pokecard-data">Type: {type} </div>
						</React.Fragment>
					) : (
						<React.Fragment>
							<h1 className="Pokecard-title">Dummy</h1>
							<div className="Pokecard-image">
								<h1>Pokemon Card</h1>
							</div>
							<div className="Pokecard-data">Dummy</div>
							<div className="Pokecard-data">Dummy</div>
						</React.Fragment>
					)}
				</div>
			</div>
		</div>
	);
};
export default Pokecard;
