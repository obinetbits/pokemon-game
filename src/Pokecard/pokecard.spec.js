import React from 'react';
import { shallow } from 'enzyme';

import Pokecard from './pokecard';

describe('Pokecard', () => {
	const baseProps = {
		wasJustClickedOn: false,
		flipCard: jest.fn(),
		uniqueId: 'so-unique',
		name: 'name',
		imageId: 'imageId',
		isCurrentName: true,
		isPrevName: false,
		isMatched: false,
		exp: 'unknown',
		type: 'unknown'
	};
	it('should render correctly in "debug" mode', () => {
		const wrapper = shallow(<Pokecard debug />);

		expect(wrapper).toMatchSnapshot();
	});
	it('should render with card closed on initialy', () => {
		const wrapper = shallow(<Pokecard {...baseProps} />);

		expect(wrapper.find('.flip-card-action')).toHaveLength(0);
		expect(wrapper).toMatchSnapshot();
	});
	it('should render with card open when clicked on', () => {
		const props = {
			...baseProps,
			wasJustClickedOn: true
		};
		const wrapper = shallow(<Pokecard {...props} />);

		expect(wrapper.find('.flip-card-action')).toHaveLength(1);
		expect(wrapper).toMatchSnapshot();
	});
	it('should render with card open when matching card has just been clicked on', () => {
		const props = {
			...baseProps,
			isPrevName: true
		};
		const wrapper = shallow(<Pokecard {...props} />);

		expect(wrapper.find('.flip-card-action')).toHaveLength(1);
		expect(wrapper).toMatchSnapshot();
	});
	it('should always render with card open after being matched', () => {
		const props = {
			...baseProps,
			isCurrentName: true,
			isMatched: true
		};
		const wrapper = shallow(<Pokecard {...props} />);

		expect(wrapper.find('.flip-card-action')).toHaveLength(1);
		expect(wrapper).toMatchSnapshot();
	});
	it('should flip card when clicked', () => {
		const props = {
			...baseProps,
			wasJustClickedOn: false
		};
		const wrapper = shallow(<Pokecard {...props} />);
		wrapper.find('.Pokecard-front').simulate('click');
		expect(baseProps.flipCard).toHaveBeenCalled();
		baseProps.flipCard.mockClear();
	});
	it('should not flip already opened card', () => {
		const props = {
			...baseProps,
			wasJustClickedOn: true
		};
		const wrapper = shallow(<Pokecard {...props} />);
		wrapper.find('.Pokecard-front').simulate('click');
		expect(baseProps.flipCard).not.toHaveBeenCalled();
	});
});
