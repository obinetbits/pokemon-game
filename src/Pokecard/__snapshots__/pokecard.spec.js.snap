// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokecard should always render with card open after being matched 1`] = `
<div
  className="flip-card flip-card-action"
>
  <div
    className="flip-card-inner"
  >
    <div
      className="Pokecard-front"
      onClick={[Function]}
    >
      <h1
        className="Pokecard-title"
      >
        Guess
      </h1>
      <div
        className="Pokecard-image"
      >
        <h1>
          Pokemon Card
        </h1>
      </div>
      <div
        className="Pokecard-data"
      >
        <span
          aria-label="Monser"
          role="img"
        >
          👾👾👾
        </span>
      </div>
      <div
        className="Pokecard-data"
      >
        <span
          aria-label="Diamonds"
          role="img"
        >
          ♦️♦️️♦️♦️
        </span>
      </div>
    </div>
    <div
      className="Pokecard-back"
    >
      <h1
        className="Pokecard-title"
      >
        name
      </h1>
      <div
        className="Pokecard-image"
      >
        <img
          alt="name"
          onError={[Function]}
          src="http://www.pokestadium.com/sprites/xy/name.gif"
        />
      </div>
      <div
        className="Pokecard-data"
      >
        Exp: 
        unknown
         
      </div>
      <div
        className="Pokecard-data"
      >
        Type: 
        unknown
         
      </div>
    </div>
  </div>
</div>
`;

exports[`Pokecard should render correctly in "debug" mode 1`] = `
<div
  className="flip-card"
>
  <div
    className="flip-card-inner"
  >
    <div
      className="Pokecard-front"
      onClick={[Function]}
    >
      <h1
        className="Pokecard-title"
      >
        Guess
      </h1>
      <div
        className="Pokecard-image"
      >
        <h1>
          Pokemon Card
        </h1>
      </div>
      <div
        className="Pokecard-data"
      >
        <span
          aria-label="Monser"
          role="img"
        >
          👾👾👾
        </span>
      </div>
      <div
        className="Pokecard-data"
      >
        <span
          aria-label="Diamonds"
          role="img"
        >
          ♦️♦️️♦️♦️
        </span>
      </div>
    </div>
    <div
      className="Pokecard-back"
    >
      <h1
        className="Pokecard-title"
      >
        Dummy
      </h1>
      <div
        className="Pokecard-image"
      >
        <h1>
          Pokemon Card
        </h1>
      </div>
      <div
        className="Pokecard-data"
      >
        Dummy
      </div>
      <div
        className="Pokecard-data"
      >
        Dummy
      </div>
    </div>
  </div>
</div>
`;

exports[`Pokecard should render with card closed on initialy 1`] = `
<div
  className="flip-card"
>
  <div
    className="flip-card-inner"
  >
    <div
      className="Pokecard-front"
      onClick={[Function]}
    >
      <h1
        className="Pokecard-title"
      >
        Guess
      </h1>
      <div
        className="Pokecard-image"
      >
        <h1>
          Pokemon Card
        </h1>
      </div>
      <div
        className="Pokecard-data"
      >
        <span
          aria-label="Monser"
          role="img"
        >
          👾👾👾
        </span>
      </div>
      <div
        className="Pokecard-data"
      >
        <span
          aria-label="Diamonds"
          role="img"
        >
          ♦️♦️️♦️♦️
        </span>
      </div>
    </div>
    <div
      className="Pokecard-back"
    >
      <h1
        className="Pokecard-title"
      >
        Dummy
      </h1>
      <div
        className="Pokecard-image"
      >
        <h1>
          Pokemon Card
        </h1>
      </div>
      <div
        className="Pokecard-data"
      >
        Dummy
      </div>
      <div
        className="Pokecard-data"
      >
        Dummy
      </div>
    </div>
  </div>
</div>
`;

exports[`Pokecard should render with card open when clicked on 1`] = `
<div
  className="flip-card flip-card-action"
>
  <div
    className="flip-card-inner"
  >
    <div
      className="Pokecard-front"
      onClick={[Function]}
    >
      <h1
        className="Pokecard-title"
      >
        Guess
      </h1>
      <div
        className="Pokecard-image"
      >
        <h1>
          Pokemon Card
        </h1>
      </div>
      <div
        className="Pokecard-data"
      >
        <span
          aria-label="Monser"
          role="img"
        >
          👾👾👾
        </span>
      </div>
      <div
        className="Pokecard-data"
      >
        <span
          aria-label="Diamonds"
          role="img"
        >
          ♦️♦️️♦️♦️
        </span>
      </div>
    </div>
    <div
      className="Pokecard-back"
    >
      <h1
        className="Pokecard-title"
      >
        name
      </h1>
      <div
        className="Pokecard-image"
      >
        <img
          alt="name"
          onError={[Function]}
          src="http://www.pokestadium.com/sprites/xy/name.gif"
        />
      </div>
      <div
        className="Pokecard-data"
      >
        Exp: 
        unknown
         
      </div>
      <div
        className="Pokecard-data"
      >
        Type: 
        unknown
         
      </div>
    </div>
  </div>
</div>
`;

exports[`Pokecard should render with card open when matching card has just been clicked on 1`] = `
<div
  className="flip-card flip-card-action"
>
  <div
    className="flip-card-inner"
  >
    <div
      className="Pokecard-front"
      onClick={[Function]}
    >
      <h1
        className="Pokecard-title"
      >
        Guess
      </h1>
      <div
        className="Pokecard-image"
      >
        <h1>
          Pokemon Card
        </h1>
      </div>
      <div
        className="Pokecard-data"
      >
        <span
          aria-label="Monser"
          role="img"
        >
          👾👾👾
        </span>
      </div>
      <div
        className="Pokecard-data"
      >
        <span
          aria-label="Diamonds"
          role="img"
        >
          ♦️♦️️♦️♦️
        </span>
      </div>
    </div>
    <div
      className="Pokecard-back"
    >
      <h1
        className="Pokecard-title"
      >
        name
      </h1>
      <div
        className="Pokecard-image"
      >
        <img
          alt="name"
          onError={[Function]}
          src="http://www.pokestadium.com/sprites/xy/name.gif"
        />
      </div>
      <div
        className="Pokecard-data"
      >
        Exp: 
        unknown
         
      </div>
      <div
        className="Pokecard-data"
      >
        Type: 
        unknown
         
      </div>
    </div>
  </div>
</div>
`;
